ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,935 @@
+    const cluster = opt?.cluster ?? 'mainnet-beta';
+    const ammProgram = new Program<Amm>(AmmIdl, PROGRAM_ID, provider);
+    const vaultProgram = new Program<Vault>(VaultIdl, VAULT_PROGRAM_ID, provider);
+
+    const [apyPda] = await PublicKey.findProgramAddress(
+      [Buffer.from(SEEDS.APY), pool.toBuffer()],
+      ammProgram.programId,
+    );
+
+    const tokenListContainer = await new TokenListProvider().resolve();
+    const tokenMap = cluster === 'devnet' ? DEVNET_COIN : tokenListContainer.filterByClusterSlug(cluster).getList();
+
+    const poolState = await getPoolState(pool, ammProgram);
+
+    const [vaultA, vaultB] = await Promise.all([
+      getVaultState(poolState.aVault, vaultProgram),
+      getVaultState(poolState.bVault, vaultProgram),
+    ]);
+
+    const tokenInfoA = tokenMap.find((token) => token.address === poolState.tokenAMint.toBase58());
+    const tokenInfoB = tokenMap.find((token) => token.address === poolState.tokenBMint.toBase58());
+    invariant(tokenInfoA, `TokenInfo ${poolState.tokenAMint.toBase58()} A not found`);
+    invariant(tokenInfoB, `TokenInfo ${poolState.tokenBMint.toBase58()} A not found`);
+
+    const accountsBuffer = await getAccountsBuffer({
+      connection: provider.connection,
+      vaultA,
+      vaultB,
+      apyPda,
+      poolState,
+    });
+    const accountsInfo = deserializeAccountsBuffer(accountsBuffer);
+
+    const depegAccounts = await getDepegAccounts(ammProgram.provider.connection);
+
+    let swapCurve;
+    if ('stable' in poolState.curveType) {
+      const { amp, depeg, tokenMultiplier } = poolState.curveType['stable'];
+      swapCurve = new StableSwap(amp.toNumber(), tokenMultiplier, depeg, depegAccounts, accountsInfo.currentTime);
+    } else {
+      swapCurve = new ConstantProductSwap();
+    }
+
+    const poolInfo = calculatePoolInfo(
+      accountsInfo.currentTime,
+      accountsInfo.poolVaultALp,
+      accountsInfo.poolVaultBLp,
+      accountsInfo.vaultALpSupply,
+      accountsInfo.vaultBLpSupply,
+      accountsInfo.poolLpSupply,
+      accountsInfo.apy,
+      swapCurve,
+      vaultA,
+      vaultB,
+    );
+
+    return new AmmImpl(
+      pool,
+      ammProgram,
+      vaultProgram,
+      apyPda,
+      [tokenInfoA, tokenInfoB],
+      poolState,
+      poolInfo,
+      vaultA,
+      vaultB,
+      accountsInfo,
+      swapCurve,
+      depegAccounts,
+      {
+        cluster,
+      },
+    );
+  }
+
+  get tokenA(): TokenInfo {
+    return this.tokenInfos[0];
+  }
+
+  get tokenB(): TokenInfo {
+    return this.tokenInfos[1];
+  }
+
+  get decimals(): number {
+    return Math.max(this.tokenA.decimals, this.tokenB.decimals);
+  }
+
+  get isStablePool(): boolean {
+    return 'stable' in this.poolState.curveType;
+  }
+
+  /**
+   * It updates the state of the pool
+   */
+  public async updateState() {
+    const poolState = await getPoolState(this.address, this.program);
```

Comment:
We can optimize this by
```
const [poolState, vaultA, vaultB] = awapt Promise.all([
   getPoolState(this.address, this.program),
   getVaultState(this.poolState.aVault, this.vaultProgram),
   getVaultState(this.poolState.bVault, this.vaultProgram)
]);
```
Since `aVault` and `bVault` will not be changed, we can reuse back the old a/bVault pubkey

Reply:
I've change the vaultA, vaultB state to be VaultImpl class. So i'll call `refreshVaultState` instead.